// @renproject/chains-solana/build/module/networks.js
import { RenNetwork, } from "@renproject/interfaces";
const isSolNetworkConfig = (x) => {
    return x.genesisHash !== undefined;
};
export const resolveNetwork = (renNetwork) => {
    let networkString = "";
    if (typeof renNetwork !== "string") {
        if (isSolNetworkConfig(renNetwork)) {
            return renNetwork;
        }
        else {
            networkString = renNetwork.name;
        }
    }
    else {
        networkString = renNetwork;
    }
    switch (networkString) {
        case RenNetwork.MainnetVDot3:
        case RenNetwork.Mainnet:
            return renMainnet;
        case RenNetwork.TestnetVDot3:
        case RenNetwork.Testnet:
            return renTestnet;
        case RenNetwork.DevnetVDot3:
            return renDevnet;
    }
    return renLocalnet;
};
export const renMainnet = {
    name: RenNetwork.MainnetVDot3,
    chain: "mainnet",
    isTestnet: false,
    chainLabel: "Mainnet",
    endpoint: "https://ren.rpcpool.com/",
    chainExplorer: "https://explorer.solana.com",
    lightnode: "https://lightnode-mainnet.herokuapp.com",
    addresses: {
        GatewayRegistry: "REGrPFKQhRneFFdUV3e9UDdzqUJyS6SKj88GdXFCRd2",
    },
    genesisHash: "5eykt4UsFv8P8NJdTREpY1vzqKqZKvdpKuc147dw2N9d",
};
export const renTestnet = {
    name: RenNetwork.TestnetVDot3,
    chain: "testnet",
    isTestnet: true,
    chainLabel: "Testnet",
    endpoint: "https://api.testnet.solana.com",
    chainExplorer: "https://explorer.solana.com",
    lightnode: "https://lightnode-testnet.herokuapp.com",
    addresses: {
        GatewayRegistry: "REGrPFKQhRneFFdUV3e9UDdzqUJyS6SKj88GdXFCRd2",
    },
    genesisHash: "4uhcVJyU9pJkvQyS88uRDiswHXSCkY3zQawwpjk2NsNY",
};
export const renDevnet = {
    name: RenNetwork.DevnetVDot3,
    chain: "devnet",
    isTestnet: true,
    chainLabel: "Devnet",
    endpoint: "https://api.devnet.solana.com",
    chainExplorer: "https://explorer.solana.com",
    lightnode: "https://lightnode-devnet.herokuapp.com",
    addresses: {
        GatewayRegistry: "REGrPFKQhRneFFdUV3e9UDdzqUJyS6SKj88GdXFCRd2",
    },
    genesisHash: "EtWTRABZaYq6iMfeYKouRu166VU2xqa1wcaWoxPkrZBG",
};
export const renLocalnet = {
    name: RenNetwork.Localnet,
    chain: "localnet",
    isTestnet: true,
    chainLabel: "",
    endpoint: "http://0.0.0.0:8899",
    lightnode: "http://0.0.0.0:5000",
    chainExplorer: "https://explorer.solana.com",
    addresses: {
        GatewayRegistry: "REGrPFKQhRneFFdUV3e9UDdzqUJyS6SKj88GdXFCRd2",
    },
    genesisHash: "",
};
